src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -240,6 +213,104 @@ private AsyncProtocolRequest GetOrCreateProtocolRequest(ref AsyncProtocolRequest
             return request;
         }
 
+        private async ValueTask<int> ReadAsyncInternal<TReadAdapter>(TReadAdapter adapter, byte[] buffer, int offset, int count)
+            where TReadAdapter : ISslReadAdapter
+        {
+            if (Interlocked.Exchange(ref _nestedRead, 1) == 1)
+            {
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(ReadAsync), "read"));
+            }
+
+            while (true)
+            {
+                int copyBytes;
+                if (_decryptedBytesCount != 0)
+                {
+                    copyBytes = CopyDecryptedData(buffer, offset, count);
+                    _nestedRead = 0;
+                    return copyBytes;
+                }
+                copyBytes = await adapter.LockAsync(buffer, offset, count);
+                try
+                {
+                    if (copyBytes > 0)
+                    {
+                        return copyBytes;
+                    }
+
+                    ResetReadBuffer();
+                    int readBytes = await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize);
+                    int payloadBytes = _sslState.GetRemainingFrameSize(_internalBuffer, _internalOffset, readBytes);
+                    if (payloadBytes < 0)
+                    {
+                        throw new IOException(SR.net_frame_read_size);
+                    }
+
+                    readBytes = await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize + payloadBytes);
+                    if (readBytes < 0)
+                    {
+                        throw new IOException(SR.net_frame_read_size);
+                    }
+
+                    // At this point, readBytes contains the size of the header plus body.
+                    // Set _decrytpedBytesOffset/Count to the current frame we have (including header)
+                    // DecryptData will decrypt in-place and modify these to point to the actual decrypted data, which may be smaller.
+                    _decryptedBytesOffset = _internalOffset;
+                    _decryptedBytesCount = readBytes;
+                    SecurityStatusPal status = _sslState.DecryptData(_internalBuffer, ref _decryptedBytesOffset, ref _decryptedBytesCount);
+
+                    // Treat the bytes we just decrypted as consumed
+                    // Note, we won't do another buffer read until the decrypted bytes are processed
+                    ConsumeBufferedBytes(readBytes);
+
+                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                    {
+                        byte[] extraBuffer = null;
+                        if (_decryptedBytesCount != 0)
```

Comment:
Can this come from the pool? Or is it handed out to user code?

Reply:
Could be pooled but its going to add complexity and it only happens if there is handshake or alert happening at the same time (super rare).  The whole "lock during handshake and read/write at the sametime" was something I wanted to look at in a single PR. I was actually avoiding the handshake code and SSLState due to the PR on the options bag etc.

I have done some very basic benching of the handshake but don't have 100% solid benchmarks for it yet. But its on the books for the future ;)

Reply:
Alert or handshake that starts while writing and you end up with part of the handshake frame during the read....

Reply:
Ok, thanks.